docs/source/tutorial_torch_generator_agent.md

```diff
@@ -0,0 +1,154 @@
+```
+    def reorder_encoder_states(self, encoder_states, indices):
+        h, c = encoder_states
+        return h[:, indices, :], c[:, indices, :]
+
+    def reorder_decoder_incremental_state(self, incr_state, indices):
+        h, c = incr_state
+        return h[:, indices, :], c[:, indices, :]
+```
+
+### Creating the encoder
+
+The encoder is straightfoward: it contains an embedding layer and an LSTM, and a forward pass through the encoder consists of passing the sequences of input tokens through both of them sequentially. The final hidden state is returned.
+
+```
+class Encoder(nn.Module):
+
+    def __init__(self, embeddings, hidden_size):
+        super().__init__()
+        _vocab_size, esz = embeddings.weight.shape
+        self.embeddings = embeddings
+        self.lstm = nn.LSTM(
+            input_size=esz, hidden_size=hidden_size, num_layers=1, batch_first=True
+        )
+
+    def forward(self, input_tokens):
+        embedded = self.embeddings(input_tokens)
+        _output, hidden = self.lstm(embedded)
+        return hidden
+```
+
+### Creating the decoder
+
+The decoder is initialized in the same way as the encoder, but now the forward pass reflects the fact that the input tokens need to be passed through the embedder and LSTM one token at a time rather than all at once. If this is the first pass through the decoder, we pass a tuple `encoder_state` to the LSTM that consists of the initial hidden and cell state, as taken from the output of the encoder. If this is a subsequent pass through the decoder, the LSTM will have given us the current values of the hidden and cell states, so we pass that back in to the LSTM, after potentially having reindexed the states with `ExampleModel().reorder_decoder_incremental_state()`.
+
+```
+class Decoder(nn.Module):
+
+    def __init__(self, embeddings, hidden_size):
+        super().__init__()
+        _vocab_size, self.esz = embeddings.weight.shape
+        self.embeddings = embeddings
+        self.lstm = nn.LSTM(
+            input_size=self.esz, hidden_size=hidden_size, num_layers=1, batch_first=True
+        )
+
+    def forward(self, input, encoder_state, incr_state=None):
+        embedded = self.embeddings(input)
+        if incr_state is None:
+            state = encoder_state
+        else:
+            state = incr_state
+        output, incr_state = self.lstm(embedded, state)
+        return output, incr_state
+```
+
+
+### Training
+
+The full code for the agent can be seen [here](https://github.com/facebookresearch/ParlAI/tree/master/parlai/agents/examples/seq2seq.py). To call training:
+
+```
+python examples/train_model.py -m examples/seq2seq \
+    -mf /tmp/example_model \
+    -t convai2 -bs 32 -eps 2 --truncate 128
+```
+
+You should get a perplexity of around 135 and a token accuracy of around 28% on the ConvAI2 validation/test set.
```

Comment:
Also add some bits about generation, using beam search, nucleus, etc.

Reply:
Makes sense - added this to the "Creating a Model" section